File: GitToolBox/src/main/java/zielu/gittoolbox/fetch/AutoFetchState.java
package zielu.gittoolbox.fetch;

import com.intellij.openapi.components.BaseComponent;
import com.intellij.openapi.project.Project;
import com.intellij.util.messages.MessageBusConnection;
import java.util.concurrent.atomic.AtomicBoolean;
import org.jetbrains.annotations.NotNull;
import zielu.gittoolbox.extension.AutoFetchAllowed;
import zielu.gittoolbox.extension.AutoFetchAllowedExtension;

public class AutoFetchState implements BaseComponent {
  private final AtomicBoolean fetchRunning = new AtomicBoolean();
  private final Project project;
  private final AutoFetchAllowedExtension extension;

  AutoFetchState(@NotNull Project project) {
    this.project = project;
    extension = new AutoFetchAllowedExtension(project);
  }

  @NotNull
  public static AutoFetchState getInstance(@NotNull Project project) {
    return project.getComponent(AutoFetchState.class);
  }

  @Override
  public void initComponent() {
    connectToMessageBus();
  }

  private void connectToMessageBus() {
    MessageBusConnection connection = project.getMessageBus().connect(project);
    connection.subscribe(AutoFetchAllowed.TOPIC, allowed -> fireStateChanged());
  }

  private void fireStateChanged() {
    project.getMessageBus().syncPublisher(AutoFetchNotifier.TOPIC).stateChanged(this);
  }

  boolean canAutoFetch() {
    return extension.isFetchAllowed() && !fetchRunning.get();
  }

  boolean fetchStart() {
    return fetchRunning.compareAndSet(false, true);
  }

  void fetchFinish() {
    fetchRunning.compareAndSet(true, false);
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/fetch/AutoFetchState.kt
package zielu.gittoolbox.fetch

import com.intellij.openapi.components.BaseComponent
import com.intellij.openapi.project.Project
import zielu.gittoolbox.extension.autofetch.AUTO_FETCH_ALLOWED_TOPIC
import zielu.gittoolbox.extension.autofetch.AutoFetchAllowed
import zielu.gittoolbox.extension.autofetch.AutoFetchAllowedExtension
import zielu.gittoolbox.extension.autofetch.AutoFetchAllowedNotifier
import zielu.gittoolbox.util.AppUtil
import java.util.concurrent.atomic.AtomicBoolean

internal class AutoFetchState(private val project: Project) : BaseComponent {
  private val extension = AutoFetchAllowedExtension(project)
  private val fetchRunning = AtomicBoolean()

  override fun initComponent() {
    val connection = project.messageBus.connect(project)
    connection.subscribe(AUTO_FETCH_ALLOWED_TOPIC, object : AutoFetchAllowedNotifier {
      override fun stateChanged(allowed: AutoFetchAllowed) {
        fireStateChanged()
      }
    })
  }

  private fun fireStateChanged() {
    project.messageBus.syncPublisher(AutoFetchNotifier.TOPIC).stateChanged(this)
  }

  fun canAutoFetch() = !fetchRunning.get() && extension.isFetchAllowed()

  fun fetchStart() = fetchRunning.compareAndSet(false, true)

  fun fetchFinish() {
    fetchRunning.compareAndSet(true, false)
  }

  companion object {
    @JvmStatic
    fun getInstance(project: Project): AutoFetchState {
      return AppUtil.getComponent(project, AutoFetchState::class.java)
    }
  }
}
